From 107cff7ea2e167852aae44598e8c12ceb098c821 Mon Sep 17 00:00:00 2001
From: Hyun Kwon <hyun.kwon@xilinx.com>
Date: Wed, 11 Apr 2018 09:36:33 -0700
Subject: [PATCH 0344/1852] drm: xlnx: fb: Prefer the current format depth over
 depth from fb helper

commit ee2174d7ef4e37b7fad67024006cb90f940f7ed6 from
https://github.com/Xilinx/linux-xlnx.git

The drm fb helper has specific preference of bpp and depth. For example,
for 32bit bpp, the depth is hard-coded to be 24. If it's not aligned
with the supported format of a drm device, it fails to initialize fbdev.
So override the depth value from fb helper with the current format
of the drm device. This will allow to initialize the fbdev with
preferred format that matches with actual format.

Signed-off-by: Hyun Kwon <hyun.kwon@xilinx.com>
Reviewed-by: Saurabh Sengar <saurabhs@xilinx.com>
Tested-by: Anil Kumar Mamidala <amamidal@xilinx.com>
Signed-off-by: Michal Simek <michal.simek@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/gpu/drm/xlnx/xlnx_fb.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/drivers/gpu/drm/xlnx/xlnx_fb.c b/drivers/gpu/drm/xlnx/xlnx_fb.c
index 81e46f8f8f3e..520b9cf271a6 100644
--- a/drivers/gpu/drm/xlnx/xlnx_fb.c
+++ b/drivers/gpu/drm/xlnx/xlnx_fb.c
@@ -27,6 +27,7 @@
 #include <drm/drm_gem_cma_helper.h>
 #include <drm/drm_gem_framebuffer_helper.h>
 
+#include "xlnx_crtc.h"
 #include "xlnx_drv.h"
 #include "xlnx_fb.h"
 
@@ -160,6 +161,8 @@ static int xlnx_fbdev_create(struct drm_fb_helper *fb_helper,
 	unsigned int bytes_per_pixel;
 	unsigned long offset;
 	struct fb_info *fbi;
+	u32 format;
+	const struct drm_format_info *info;
 	size_t bytes;
 	int ret;
 
@@ -182,6 +185,12 @@ static int xlnx_fbdev_create(struct drm_fb_helper *fb_helper,
 		goto err_drm_gem_cma_free_object;
 	}
 
+	/* Override the depth given by fb helper with current format value */
+	format = xlnx_get_format(drm);
+	info = drm_format_info(format);
+	if (size->surface_bpp == info->cpp[0] * 8)
+		size->surface_depth = info->depth;
+
 	fbdev->fb = xlnx_fb_gem_fbdev_fb_create(drm, size, fbdev->align,
 						&obj->base, &xlnx_fb_funcs);
 	if (IS_ERR(fbdev->fb)) {
-- 
2.31.1

